Mathlib/Tactic/DeriveEncodable.lean

```diff
@@ -0,0 +1,363 @@
+        let mut ctorArgs : Array Term := #[]
+        let mut patternArgs2 : Array Term := #[]
+        let mut patternArgs3 : Array Term := #[]
+        for _ in [:indVal.numParams] do
+          ctorArgs := ctorArgs.push (← `(_))
+        for i in [:ctorInfo.numFields] do
+          let a := mkIdent (← mkFreshUserName `a)
+          let x := xs[ctorInfo.numParams + i]!
+          let xTy ← inferType x
+          let recName? := ctx.typeInfos.findIdx? (xTy.isAppOf ·.name) |>.map (fromSNames[·]!)
+          if let some recName := recName? then
+            patternArgs := patternArgs.push a
+            discrs := discrs.push <| ← `(matchDiscr| $(mkIdent recName) $a)
+          else
+            patternArgs := patternArgs.push <| ← ``(S.nat $a)
+            discrs := discrs.push <| ← `(matchDiscr| $(mkCIdent ``Encodable.decode) $a)
+          ctorArgs := ctorArgs.push a
+          patternArgs2 := patternArgs2.push <| ← ``(some $a)
+          patternArgs3 := patternArgs3.push <| ← `(_)
+        let pattern ← patternArgs.foldrM (init := ← ``(S.nat 0)) fun arg acc => ``(S.cons $arg $acc)
+        let pattern ← ``(S.cons (S.nat $(quote ctorInfo.cidx)) $pattern)
+        -- Note: this is where we could try to handle indexed types.
+        -- The idea would be to use DecidableEq to test the computed index against the expected
+        -- index and then rewrite.
+        let res ← ``(some <| @$(mkIdent ctorName):ident $ctorArgs:term*)
+        if discrs.isEmpty then
+          `(matchAltExpr| | $pattern:term => $res)
+        else
+          let rhs : Term ← `(
+              match $[$discrs],* with
+              | $[$patternArgs2],* => $res
+              | $[$patternArgs3],* => none
+            )
+          `(matchAltExpr| | $pattern:term => $rhs)
+    alts := alts.push <| ← `(matchAltExpr| | _ => none)
+    return alts
+
+/-- Constructs a function from `S` to the inductive type. -/
+def mkFromSFuns (ctx : Deriving.Context) (fromSFunNames : Array Name) :
+    TermElabM (TSyntax `command) := do
+  let mut res : Array (TSyntax `command) := #[]
+  for i in [:fromSFunNames.size] do
+    let fromNatFnName := fromSFunNames[i]!
+    let indVal := ctx.typeInfos[i]!
+    let header ← mkHeader ``Encodable 1 indVal
+    let body ← mkFromSMatch ctx indVal fromSFunNames
+    -- Last binder is for the target
+    let binders := header.binders[0:header.binders.size - 1]
+    res := res.push <| ← `(
+      private def $(mkIdent fromNatFnName):ident $binders:bracketedBinder* :
+        $(mkCIdent ``S) → Option $header.targetType := $body:term
+      )
+  `(command| mutual $[$res:command]* end)
+
+/-!
+Constructing the proofs that the `fromS` functions are left inverses of the `toS` functions.
+-/
+
+/--
+Constructs a proof that the functions created by `mkFromSFuns` are left inverses
+of the ones created by `mkToSFuns`.
+-/
+def mkInjThms (ctx : Deriving.Context) (toSFunNames fromSFunNames : Array Name) :
```

Comment:
I don't understand why you chose to mark everything as `private` but not this and `mkFromSFuns `. Is this on purpose?

Reply:
It's just an oversight. Fixed it.